compiler/damlc/lib/DA/Cli/Damlc/DependencyDb.hs

```diff
@@ -175,24 +186,55 @@ dalfFileName bs fp = do
     let pkgId = T.unpack $ LF.unPackageId $ LF.dalfPackageId decodedDalfPkg
     pure $ pkgId </> takeFileName fp
 
-installDataDepDalf :: FilePath -> FilePath -> IO ()
-installDataDepDalf depsDir fp = do
-    bs <- BS.readFile fp
+installDataDepDalf :: Bool -> FilePath -> FilePath -> BS.ByteString -> IO ()
+installDataDepDalf isMain depsDir fp bs = do
     fileName <- dalfFileName bs fp
     let targetFp = depsDir </> fileName
     let targetDir = takeDirectory targetFp
-    markDirWith mainMarker targetDir
+    unlessM (doesDirectoryExist targetDir) $ write targetFp $ BSL.fromStrict bs
+    -- if the directory exists, the dalf got already installed as a dependency
+    when isMain $ markDirWith mainMarker targetDir
     markDirWith dataDepMarker targetDir
-    copy fp targetFp
+
+data DataDeps = DataDeps
+    { dataDepsDars :: [FilePath]
+    , dataDepsDalfs :: [FilePath]
+    , dataDepsPkgIds :: [LF.PackageId]
+    }
+
+readDataDeps :: [String] -> IO DataDeps
+readDataDeps fpOrIds = do
+    pkgIds <- forM pkgIds0 validatePkgId
+    pure $ DataDeps {dataDepsDars = dars, dataDepsDalfs = dalfs, dataDepsPkgIds = pkgIds}
+  where
+    (dars, rest) = partition ("dar" `isExtensionOf`) fpOrIds
+    (dalfs, pkgIds0) = partition ("dalf" `isExtensionOf`) rest
+
+-- | A check that no bad package ID's are present in the data-dependency section of daml.yaml.
+validatePkgId :: String -> IO LF.PackageId
+validatePkgId pkgId = do
+    unless ((Base64.isBase64 $ T.pack pkgId) && all isLower (filter isAlpha pkgId)) $
+        fail $ "Invalid package ID dependency in daml.yaml: " <> pkgId
+    pure $ LF.PackageId $ T.pack pkgId
```

Comment:
This isn't right ... package ids are lowercase hex strings (with length 64), like @cocreature said, not base64. This `validatePkgId` will accept strings like `"helloworld"` when it should reject them.

Reply:
ah no, I must have misread the comment. I thought the request was explicitly to check for base64. I'll fix it in the next PR.